### ISSUE TITLE
OpenMetrics Format

### ISSUE THREAD

**Issue by eaglerainbow** [promregator/promregator#214]
## Summary / Problem Statement
Shall Promregator also support the [OpenMetrics Format](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md)?

**Comment by checketts** [comment]
Yes and No.  ;)

As in the v2 endpoint I propose in #208 it should just passthrough the Prometheus call to the underlying service. So if that service exposed OpenMetrics, they ca be used and Promregator doesn't impact that.

**Comment by braunsonm** [comment]
Could this issue be revisited? Since Prometheus itself adds the headers for accepting OpenMetrics format, it should be possible in Promregator to add these headers and expose the results on the /metrics.

Otherwise does Promregator at least pass through all Prometheus headers when it does a scrape on behalf of Prometheus?

**Comment by braunsonm** [comment]
It would be great if Promregator could at least, as a mitigation, pass through the `Accept` header that Prometheus sends. This would insure the upstream server (Spring in our case) returns the correct format.

Assuming Promregator is just showing you the body of the response without any other changes.

**Comment by eaglerainbow** [comment]
> It would be great if Promregator could at least, as a mitigation, pass through the Accept header that Prometheus sends.

This shouldn't be the big problem.

> Assuming Promregator is just showing you the body of the response without any other changes.

Unfortunately, this assumption (as of now) does not hold true 😞 . It is one option to make that happen (btw, the same one checketts is heading for).

**Comment by braunsonm** [comment]
Oh yea actually I didn't think that through, wouldn't work in our case as we rely on Promregator to add some labels about the application. Those could be added in the `/discovery` scrape configs though using relabel rules.

Anyways, eagerly tracking this issue as we would like to start ingesting exemplars and this is the blocker right now for us.

**Comment by checketts** [comment]
I'll look into the header passthrough. Very good point there.

**Comment by checketts** [comment]
I just verified that my technique is not passing headers through. I'll fix that and rebase my PR

**Comment by eaglerainbow** [comment]
Internal remark:
* official specification: https://github.com/OpenObservability/OpenMetrics/blob/1386544931307dff279688f332890c31b6c5de36/specification/OpenMetrics.md
* see also how spring-boot did it: https://github.com/spring-projects/spring-boot/commit/11b4a19dee3a54fbb5df16fbcf5bd6413968a5b2

**Comment by eaglerainbow** [comment]
Release of [1.0.0-rc4 is there](https://github.com/promregator/promregator/releases/tag/v1.0.0-rc4)

**Comment by braunsonm** [comment]
@eaglerainbow your proposal addresses all my concerns, looks great to me.

**Comment by checketts** [comment]
> [@checketts](https://github.com/checketts): This approach also made the "Parser" class, which you apparently dislike a lot 😁, obsolete. Hence it is also another step towards "SimpleProm".

🥹

Sorry the PromClient breaking changes were the catalyst though.

**Comment by eaglerainbow** [comment]
> Sorry the PromClient breaking changes were the catalyst though.

No worries! Your approach is giving the right direction. It's only that for me they are a little too progressive to apply them at once...

**Comment by eaglerainbow** [comment]
@checketts : In a real life test it turned out that the OpenFormat's two Metric Sets cannot just be concatenated like this: There is the `# EOF` endmarker on which Prometheus is very picky about where it sits. 

If anyone wants to see what "joy" this creates, you are referred to https://github.com/promregator/promregator/blob/1e1cfff46f6af452ca1682a9311719c033272466/src/main/java/org/cloudfoundry/promregator/rewrite/MetricSetMerger.java.

**Comment by eaglerainbow** [comment]
fwiw: With [1.0.0-rc6](https://github.com/promregator/promregator/releases/tag/v1.0.0-rc6) I was able the first time to run on a OpenMetrics-only environment end-2-end. Prometheus was a 2.41.0. 
I want to keep it running for a while to also see, if it causes unknown operational issues.

@braunsonm : If you have time, you might also want to have a look...

**Comment by braunsonm** [comment]
Sure I can try it out. Is there any breaking changes that constitute a 1.0 release I should be aware of? Or any preliminary release notes?

**Comment by checketts** [comment]
@eaglerainbow  Interesting choices for their format. Cheers on the the thoroughness! (And I confess my laziness in just passing it through in my implementation)

**Comment by eaglerainbow** [comment]
> Sure I can try it out. Is there any breaking changes that constitute a 1.0 release I should be aware of? Or any preliminary release notes?

@braunsonm 1.0.0 is different in many regards compared to 0.x.y.  The most important differences I tried to describe at https://github.com/promregator/promregator/blob/1e1cfff46f6af452ca1682a9311719c033272466/README.md#differences-on-major-versions . 
I also started to write something like a "migration guide" from 0.x.y to 1.0.0. You can find it [here](https://github.com/promregator/promregator/blob/1e1cfff46f6af452ca1682a9311719c033272466/docs/upgradev1.md).

Feedback on both documents are welcome btw (perhaps then better as PR or in an own issue - otherwise this issue here risks to become blurry)!

**Comment by eaglerainbow** [comment]
> And I confess my laziness in just passing it through in my implementation

There is nothing to confess :) If your version runs without it in your environment, then it does its job 😁 
BTW: IIRC your implementation also got rid of the "additional metrics" as part of the main single endpoint request, but you add those to the promregator's metrics request itself. Then you wouldn't into that issue so much. However, I also see some more drawbacks with the approach, which is why I get in conflict with these format intricacies...

**Comment by eaglerainbow** [comment]
@braunsonm  Yes, please - do so. I can't reproduce it right now, so I would need a little more context...
Please mention in that issue
a) that you are testing 1.0.0-rc*something*,
b) which passwords you have entered where (not the values obviously, but stating *that* a password was provided). 

Also the stacktrace might be helpful.
I'll have a look then.

Thanks!

**Comment by eaglerainbow** [comment]
@braunsonm I just released [version 1.0.0](https://github.com/promregator/promregator/releases/tag/v1.0.0), which mainly is your 1.0.0-rc7. Have found some time on testing in the meantime?

**Comment by eaglerainbow** [comment]
I wouldn't vouch for it, but given that we have changed the [MetricFetcher to return a FetchResult, which mainly only contains an (arbitrary) string](https://github.com/promregator/promregator/blob/7c0fd0dbc3cd1ffcecf420787c51cd88ec2c4df8/src/main/java/org/cloudfoundry/promregator/fetcher/CFMetricsFetcher.java#L168) and given that we only ["add some more metrics" along the way](https://github.com/promregator/promregator/blob/7c0fd0dbc3cd1ffcecf420787c51cd88ec2c4df8/src/main/java/org/cloudfoundry/promregator/endpoint/SingleTargetMetricsEndpoint.java#L186) (in a somewhat "ugly fashion", though), I wouldn't know why it shouldn't.
I'd simply say: Give it a try - and report back here on your findings 😁

**Comment by dmarmugi** [comment]
Running into a bit of an issue with this on the openmetrics-text version matching in CFMetricsFetcher:
```java
	private static final Pattern CONTENT_TYPE_OPENMETRIC_100 = Pattern.compile("^application/openmetrics-text; *version=1.[0-9]++.[0-9]++; *charset=utf-8");
```
Some libraries (looking at you, [client_python](https://github.com/prometheus/client_python/blob/master/prometheus_client/openmetrics/exposition.py#L6) are still responding with `version=0.0.1`. (There should probably be an issue open on that library as version=1.0.0 has been published, but...) Prometheus server currently doesn't validate on the version and scrapes it happily. (FWIW, there seem to be a whole raft of issues similar to this: prometheus/prometheus#9430, prometheus/client_java#702, etc.)

As a pass-through on /discovery, I'd suggest the below, which also escapes the `.` to match the literal instead of a wildcard
```java
"^application/openmetrics-text; *version=[01]\\.[0-9]++\\.[0-9]++; *charset=utf-8"
```

**Comment by eaglerainbow** [comment]
@dmarmugi I see there multiple issues reported in your comment. Let's tackle them one by one:

## Dots are not Dots in a pattern (but must be escaped)
Correct, that's a bug. Fixed with 57e40450659a47afc0af2b2d39eb5f0ff444451e and 3e574238c65a48a07095a71e120d88fdce7a502c. The latter will be included in 1.0.1.

## version 0.0.1 vs. 1.0.0

😢 This "attempt" by Prometheus to "make it look like it's compatible, but in fact it's not" is getting more and more excrescences. It almost looks like that they tried to get "a half-pregnant only" with it. Put on top that I don't like these implicit "I am guessing what you have meant"-pieces of logic in communication adapters like promregator is. That is why I don't want to just soften the regular expression, but detect the situation properly and make users aware that something's fishy here. They should go exactly the same way as you already have done for python and report to the client's libraries, that they are wrong.
In that spirit, I have created a corresponding check that writes a warning message into the logs, plus a [GH wiki page](https://github.com/promregator/promregator/wiki/Invalid-Version-In-Content-Type-of-OpenMetrics-Endpoints) which explains the issue and what you can do to prevent log flooding.
To further contribute to the "1 and 0" confusion: This "feature" will be available with promregator 1.0.1.

**Comment by eaglerainbow** [comment]
Promregator [1.0.1 is released](https://github.com/promregator/promregator/releases/tag/v1.0.1).
